Move ToServices translation to Cilium operator

nebril opened the pull request:
This PR is basically #6509 rebased on master

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7393)
<!-- Reviewable:end -->

tgraf committed:
operator: Listen to services changes even if service synchronization is disabled

This allows to use the ServiceCache for CNP translation.

Signed-off-by: Thomas Graf <thomas@cilium.io>

raybejjani commented on pkg/policy/api/egress.go:
```diff
@@ -204,12 +204,24 @@ func (e *EgressRule) CreateDerivative() (*EgressRule, error) {
-		if len(cidrSet) == 0 {
-			return &EgressRule{}, nil
+		if len(cidrSet) > 0 {
+			newRule.ToCIDRSet = append(e.ToCIDRSet, cidrSet...)
```
The append will be a no-op if cidrSet is empty, so we can call it unconditionally. The same thing is true for the handling of ToServices below.

raybejjani commented on pkg/policy/api/egress.go:
```diff
@@ -204,12 +204,24 @@ func (e *EgressRule) CreateDerivative() (*EgressRule, error) {
 	newRule.ToGroups = nil
+
+	for _, svc := range e.ToServices {
+		cidrSet := svc.GetCidrSet()
```
Is this missing an error return from GetCidrSet?

raybejjani commented on pkg/policy/api/service.go:
```diff
@@ -25,14 +62,106 @@ type Service struct {
+	}
+
+	endpointSelector := EndpointSelector(k.Selector)
+	endpointSelector.SyncRequirementsWithLabelSelector()
```
I'm not sure if it's simpler to declare ServiceSelector with
```
type ServiceSelector {
    EndpointSelector
}
```
so we can avoid this cast. It does mean that all the functions of EndpointSelector would be exported.

raybejjani reviewed (commented):
I'm still trying to fully understand this but but I think it's good. I'll keep going through it.

Is it correct that toGroups and toServices are handled separately, with separate derivative CNPs created?
Edit: No, they are both handled in `createDerivativeCNP`

raybejjani commented on operator/cnp_event.go:
```diff
@@ -59,7 +79,21 @@ func enableCNPWatcher() error {
 						}
 
-						groups.UpdateDerivativeCNPIfNeeded(newCNP.CiliumNetworkPolicy, oldCNP.CiliumNetworkPolicy)
+						if !newCNP.RequiresDerivative() {
```
The old `UpdateDerivativeCNPIfNeeded` function would also update the local copy via `addDerivativeCNP`, similar to the `AddFunc` callback just above. The code expanded here deletes old CNPs if they are updated and the new copy doesn't need a derivative. It never handles the case where newCNP changes something but still requires a derivative.

nebril replied:
Seems to me like `addDerivativeCNP` is always called in controller update if `newCNP.RequiresDerivative` is true

raybejjani commented on pkg/policy/api/egress.go:
```diff
@@ -204,12 +204,24 @@ func (e *EgressRule) CreateDerivative() (*EgressRule, error) {
 	newRule.ToGroups = nil
+
+	for _, svc := range e.ToServices {
+		cidrSet := svc.GetCidrSet()
```
I noticed that `svc.GetCidrSet` and `groups.GetCidrSet` use different code despite `svc.GetCidrSet` iterating over a list of providers (implying it can handle different types). I'm not sure where to put it, but a comment pointing this out might help to understand the code later.

raybejjani replied:
Oh, actually they are completely different... nevermind :/

raybejjani commented on operator/cnp_derivative.go:
```diff
@@ -0,0 +1,240 @@
+			derivativeCNP.Specs[i] = rule
+			continue
+		}
+		newRule, err := rule.CreateDerivative()
```
One of the commits is titled "Move CNP specific handling into operator". I wonder if we should also move `CreateDerivative` to this file. It seems to only call `CreateDerivative` on EgressRule and then just iterates through modifying public fields in `api.Rule` and `api.EgressRule` for both ToGroups and ToServices. This way all derivative code is in the operator (I don't think cilium ever checks whether something is a derivative).

nebril replied:
I think it's better to leave derivative creation in the pkg dir, this way it can be reused elsewhere if needed.

raybejjani commented on daemon/k8s_watcher.go:
```diff
@@ -996,51 +953,10 @@ func (d *Daemon) k8sServiceHandler() {
 				scopedLog.WithError(err).Error("Unable to add/update service to implement k8s event")
```
Could you add a comment to this section explaining that we manage the lbmap entries here (I think?), including updates, but that we create toCIDR equivalents in cilium-operator. I think it will be really unclear where the data is coming from in some cases.

nebril committed:
operator: Move CNP specific handling into operator

The pkg/policy/groups package was polluted by Kubernetes specific CNP
handling.
Move all of this into operator/ so it can be reused for services
translation.

Signed-off-by: Maciej Kwiek <maciej@covalent.io>

nebril committed:
operator: Move ToServices translations into operator

The translation of ToServices rules to the corresponding service backend
CIDRs
can be done in the operator similar to how ToGroups are being resolved.

Remove all of the rule translation code from the agent itself and avoid
triggering a full policy recalculation on every service update when
ToServices
rules are in effect.

Signed-off-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: Maciej Kwiek <maciej@covalent.io>

aanm commented on operator/cnp_derivative.go:
```diff
@@ -12,25 +12,37 @@
 	cilium_v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/logging/logfields"
 	"github.com/cilium/cilium/pkg/policy/api"
+	_ "github.com/cilium/cilium/pkg/policy/groups"
```
please write a comment stating why this is an empty import.

aanm commented on operator/cnp_derivative.go:
```diff
@@ -12,25 +12,37 @@
 // See the License for the specific language governing permissions and
```
Copyright year should be 2018-2019

aanm commented on operator/cnp_event.go:
```diff
@@ -48,7 +53,20 @@ func enableCNPWatcher() error {
+					controllerManager.UpdateController(fmt.Sprintf("add-derivative-cnp-%s", cnp.ObjectMeta.Name),
+						controller.ControllerParams{
+							DoFunc: func(ctx context.Context) error {
+								return addDerivativeCNP(cnp.CiliumNetworkPolicy)
```
Why is this inside a controller and not a serializer queue? Having it as controllers could cause issues concurrently for a CNP with the same name. A serializer queue will make sure an event is processed before going to the next one.

nebril replied:
Good point!

aanm commented on operator/cnp_event.go:
```diff
@@ -91,11 +132,50 @@ func enableCNPWatcher() error {
+var cnpCache = cnpCacheMap{}
+
+type cnpCacheMap struct {
+	sync.Map
```
Didn't we have problems in the past with sync.Map that had pointers as values?

aanm commented on operator/cnp_event.go:
```diff
@@ -91,11 +132,50 @@ func enableCNPWatcher() error {
+						addDerivativeCNP(cnp)
+					}(cnp)
+				}
 				return nil
```
this controller returns without making sure that all go routines started on line 147 have stopped

nebril replied:
These goroutines will run for as long as operator main function runs, won't they? Or are you concerned with controller running for more than `RunInterval` and creating goroutines second time? I think this is a valid concern, will add a wait before return.

nebril replied:
Agreed, will add waitgroup here.

aanm commented on operator/cnp_event.go:
```diff
@@ -91,11 +132,50 @@ func enableCNPWatcher() error {
+					sem <- true
+					go func(cnp *cilium_v2.CiliumNetworkPolicy) {
+						defer func() { <-sem }()
+						addDerivativeCNP(cnp)
```
100 CNP updates in a single burst will probably DDoS a kube-apiserver. I don't see the point of doing this concurrently, as these policies are eventually going to be updated with the latest information from provider in k8s. The "eventually" is the keyword here, what makes it better to do it concurrently?

nebril replied:
This is rate limited by `sem` channel with `maxConcurrentUpdates` size.

aanm commented on operator/cnp_event.go:
```diff
@@ -91,11 +132,50 @@ func enableCNPWatcher() error {
+					sem <- true
+					go func(cnp *cilium_v2.CiliumNetworkPolicy) {
+						defer func() { <-sem }()
+						addDerivativeCNP(cnp)
```
Are we blindly creating the derivative of a CNP or are we only updating the derivative CNP based on the difference in the information from providers?

nebril replied:
At this point we are updating derivatives just in case provider information has changed (it's mostly for aws `toGroups`).

aanm commented on operator/cnp_services.go:
```diff
@@ -0,0 +1,46 @@
+// Copyright 2018 Authors of Cilium
```
year

aanm commented on operator/cnp_services.go:
```diff
@@ -0,0 +1,46 @@
+}
+
+func getIPs(svc *api.Service) []net.IP {
+	result := []net.IP{}
```
`var result []net.IP`

aanm commented on pkg/policy/api/service.go:
```diff
@@ -25,14 +62,106 @@ type Service struct {
+	registeredProviders.mutex.RLock()
+	defer registeredProviders.mutex.RUnlock()
+
+	ipList := []net.IP{}
```
`var ipList []net.IP`

aanm commented on pkg/policy/api/service_test.go:
```diff
@@ -0,0 +1,165 @@
+// Copyright 2018 Authors of Cilium
```
2019

aanm commented on test/k8sT/manifests/external_service.yaml:
```diff
@@ -5,8 +5,8 @@ metadata:
     zgroup: external
     test: external
+  namespace: default
 spec:
```
Won't this create a clusterIP? Is this the intention?

nebril replied:
I think it will not, but it's leftover from my debugging. Will remove
